<!DOCTYPE html>
<html>
<head>
  <meta http-equiv='content-type' value='text/html;charset=utf8'>
  <meta name='generator' value='Ronn/v0.7.3 (http://github.com/rtomayko/ronn/tree/0.7.3)'>
  <title>maf_tile(1) - synthesize an alignment for a given region</title>
  <style type='text/css' media='all'>
  /* style: man */
  body#manpage {margin:0}
  .mp {max-width:100ex;padding:0 9ex 1ex 4ex}
  .mp p,.mp pre,.mp ul,.mp ol,.mp dl {margin:0 0 20px 0}
  .mp h2 {margin:10px 0 0 0}
  .mp > p,.mp > pre,.mp > ul,.mp > ol,.mp > dl {margin-left:8ex}
  .mp h3 {margin:0 0 0 4ex}
  .mp dt {margin:0;clear:left}
  .mp dt.flush {float:left;width:8ex}
  .mp dd {margin:0 0 0 9ex}
  .mp h1,.mp h2,.mp h3,.mp h4 {clear:left}
  .mp pre {margin-bottom:20px}
  .mp pre+h2,.mp pre+h3 {margin-top:22px}
  .mp h2+pre,.mp h3+pre {margin-top:5px}
  .mp img {display:block;margin:auto}
  .mp h1.man-title {display:none}
  .mp,.mp code,.mp pre,.mp tt,.mp kbd,.mp samp,.mp h3,.mp h4 {font-family:monospace;font-size:14px;line-height:1.42857142857143}
  .mp h2 {font-size:16px;line-height:1.25}
  .mp h1 {font-size:20px;line-height:2}
  .mp {text-align:justify;background:#fff}
  .mp,.mp code,.mp pre,.mp pre code,.mp tt,.mp kbd,.mp samp {color:#131211}
  .mp h1,.mp h2,.mp h3,.mp h4 {color:#030201}
  .mp u {text-decoration:underline}
  .mp code,.mp strong,.mp b {font-weight:bold;color:#131211}
  .mp em,.mp var {font-style:italic;color:#232221;text-decoration:none}
  .mp a,.mp a:link,.mp a:hover,.mp a code,.mp a pre,.mp a tt,.mp a kbd,.mp a samp {color:#0000ff}
  .mp b.man-ref {font-weight:normal;color:#434241}
  .mp pre {padding:0 4ex}
  .mp pre code {font-weight:normal;color:#434241}
  .mp h2+pre,h3+pre {padding-left:0}
  ol.man-decor,ol.man-decor li {margin:3px 0 10px 0;padding:0;float:left;width:33%;list-style-type:none;text-transform:uppercase;color:#999;letter-spacing:1px}
  ol.man-decor {width:100%}
  ol.man-decor li.tl {text-align:left}
  ol.man-decor li.tc {text-align:center;letter-spacing:4px}
  ol.man-decor li.tr {text-align:right;float:right}
  </style>
  <style type='text/css' media='all'>
  /* style: toc */
  .man-navigation {display:block !important;position:fixed;top:0;left:113ex;height:100%;width:100%;padding:48px 0 0 0;border-left:1px solid #dbdbdb;background:#eee}
  .man-navigation a,.man-navigation a:hover,.man-navigation a:link,.man-navigation a:visited {display:block;margin:0;padding:5px 2px 5px 30px;color:#999;text-decoration:none}
  .man-navigation a:hover {color:#111;text-decoration:underline}
  </style>
</head>
<!--
  The following styles are deprecated and will be removed at some point:
  div#man, div#man ol.man, div#man ol.head, div#man ol.man.

  The .man-page, .man-decor, .man-head, .man-foot, .man-title, and
  .man-navigation should be used instead.
-->
<body id='manpage'>
  <div class='mp' id='man'>

  <div class='man-navigation' style='display:none'>
    <a href="#NAME">NAME</a>
    <a href="#SYNOPSIS">SYNOPSIS</a>
    <a href="#DESCRIPTION">DESCRIPTION</a>
    <a href="#OPTIONS">OPTIONS</a>
    <a href="#EXAMPLES">EXAMPLES</a>
    <a href="#FILES">FILES</a>
    <a href="#ENVIRONMENT">ENVIRONMENT</a>
    <a href="#COPYRIGHT">COPYRIGHT</a>
    <a href="#SEE-ALSO">SEE ALSO</a>
  </div>

  <ol class='man-decor man-head man head'>
    <li class='tl'>maf_tile(1)</li>
    <li class='tc'>BioRuby Manual</li>
    <li class='tr'>maf_tile(1)</li>
  </ol>

  <h2 id="NAME">NAME</h2>
<p class="man-name">
  <code>maf_tile</code> - <span class="man-whatis">synthesize an alignment for a given region</span>
</p>

<h2 id="SYNOPSIS">SYNOPSIS</h2>

<p><code>maf_tile</code> [<var>options</var>] -i [SEQ:]BEGIN:END [-s SPECIES[:NAME] ...] <var>maf</var> [index]</p>

<p><code>maf_tile</code> [<var>options</var>] --bed BED -o BASE [-s SPECIES[:NAME] ...] <var>maf</var> [index]</p>

<h2 id="DESCRIPTION">DESCRIPTION</h2>

<p><strong>maf_tile</strong> takes a MAF file, with optional index, or directory of
indexed MAF files, extracts alignment blocks overlapping the given
genomic interval, and constructs a single alignment block covering the
entire interval for the specified species. Optionally, any gaps in
coverage of the MAF file's reference sequence can be filled in from a
FASTA sequence file.</p>

<p>If a single interval is specified, the output will be written to
stdout in FASTA format. If a directory of MAF files is supplied as the
<var>maf</var> parameter, the interval must include the sequence identifier in
the form <code>sequence:begin:end</code>. If the <code>--output-base</code> option is
specified, <code>_&lt;begin>:&lt;end>.fa</code> will be appended to the given <base />
parameter and used to construct the output path. If a BED file is
specified with <code>--bed</code>, <code>--output-base</code> is also required.</p>

<p>Species can be renamed for output by specifying them as SPECIES:NAME;
the first component will be used to select the species from the MAF
file, and the second will be used in the FASTA description line for
output.</p>

<h2 id="OPTIONS">OPTIONS</h2>

<dl>
<dt><code>-r</code>, <code>--reference SEQ</code></dt><dd><p>The FASTA reference sequence file given, which may be gzipped, will
be used to fill in any gaps between alignment blocks.</p></dd>
<dt><code>-i</code>, <code>--interval BEGIN:END</code></dt><dd><p>The given zero-based genomic interval will be used to select
alignment blocks from the MAF file.</p></dd>
<dt><code>-s</code>, <code>--species SPECIES[:NAME]</code></dt><dd><p>The given species will be selected for output. If given as
<code>species:name</code>, it will appear in the FASTA output as <var>name</var>.</p></dd>
<dt><code>-b</code>, <code>--bed BED</code></dt><dd><p>The given BED file will be used to provide a list of intervals to
process. If present, <code>--interval</code> will be ignored and
<code>--output-base</code> must be given as well.</p></dd>
<dt><code>-o</code>, <code>--output-base BASE</code></dt><dd><p>The given path will be used as the base name for output files, as
described above.</p></dd>
</dl>


<h2 id="EXAMPLES">EXAMPLES</h2>

<p>Generate an alignment of the <code>hg19</code>, <code>petMar1</code>, and <code>ornAna1</code>
sequences from <code>chrY.maf</code> over the interval 14400 to 15000 on the
reference sequence of the MAF file. Fills in gaps from
<code>chrY.refseq.fa.gz</code>. Writes FASTA output to stdout.</p>

<pre><code>$ maf_tile --reference ~/maf/chrY.refseq.fa.gz \
  --interval 14400:15000 \
  -s hg19:human -s petMar1 -s ornAna1 \
  chrY.maf chrY.kct
&gt;human
GGGTGACGAAAAGAGCCGA-----[...]
&gt;petMar1
gagtgccggggagtgccggggagt[...]
&gt;ornAna1
AGGGATCTGGGAATTCTGG-----[...]
</code></pre>

<p>Write out a FASTA file for each interval in the given BED file,
prefixed with <code>/tmp/mm8</code>, and without filling in data from a reference
sequence:</p>

<pre><code>$ maf_tile --bed /tmp/mm8.bed --output-base /tmp/mm8 \
  -s mm8:mouse -s rn4:rat -s hg18:human \
  mm8_chr7_tiny.maf mm8_chr7_tiny.kct
</code></pre>

<h2 id="FILES">FILES</h2>

<p>The output is generated in FASTA format, with one sequence per
species.</p>

<p>The <var>maf</var> parameter must specify either a Multiple Alignment Format
(MAF) file or a directory of such files, with indexes.</p>

<p>The <var>index</var> must be a MAF index built with <a class="man-ref" href="maf_index.1.html">maf_index<span class="s">(1)</span></a>. This can be
omitted if a single MAF file is given as a parameter, but in this case
the entire file will be parsed to build a temporary index. For large
files which will be reused, this is not advisable.</p>

<p>If <code>--bed</code> <var>bed</var> is specified, its argument must be a BED file. Only
the second and third columns will be used, to specify the zero-based
start and end positions of intervals.</p>

<h2 id="ENVIRONMENT">ENVIRONMENT</h2>

<p><code>maf_tile</code> is a Ruby program and relies on ordinary Ruby environment
variables.</p>

<h2 id="COPYRIGHT">COPYRIGHT</h2>

<p><code>maf_tile</code> is copyright (C) 2012 Clayton Wheeler.</p>

<h2 id="SEE-ALSO">SEE ALSO</h2>

<p><a class="man-ref" href="maf_index.1.html">maf_index<span class="s">(1)</span></a>, <span class="man-ref">ruby<span class="s">(1)</span></span></p>

<ul>
<li><a href="https://github.com/csw/bioruby-maf/" data-bare-link="true">https://github.com/csw/bioruby-maf/</a></li>
</ul>



  <ol class='man-decor man-foot man foot'>
    <li class='tl'>BioRuby</li>
    <li class='tc'>July 2012</li>
    <li class='tr'>maf_tile(1)</li>
  </ol>

  </div>
</body>
</html>
