
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>BioRuby MAF blog</title>
  <meta name="author" content="Clayton Wheeler">

  
  <meta name="description" content="It&#8217;s the end of my first week of actual development work, and I&#8217;m
happy with my progress. My first milestone, for batch MAF
conversion to &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://csw.github.com/bioruby-maf">
  <link href="/bioruby-maf/favicon.png" rel="icon">
  <link href="/bioruby-maf/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/bioruby-maf/javascripts/modernizr-2.0.js"></script>
  <script src="/bioruby-maf/javascripts/ender.js"></script>
  <script src="/bioruby-maf/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/bioruby-maf/atom.xml" rel="alternate" title="BioRuby MAF blog" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/bioruby-maf/">BioRuby MAF blog</a></h1>
  
    <h2>Multiple Alignment Format support for BioRuby and bio-alignment</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/bioruby-maf/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:csw.github.com/bioruby-maf" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/bioruby-maf/">Blog</a></li>
  <li><a href="/bioruby-maf/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/bioruby-maf/blog/2012/05/25/first_milestone/">First Milestone</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2012-05-25T14:13:00-04:00" pubdate data-updated="true">May 25<span>th</span>, 2012</time>
        
         | <a href="/bioruby-maf/blog/2012/05/25/first_milestone/#disqus_thread">Comments</a>
        
      </p>
    
  </header>


  <div class="entry-content"><p>It&#8217;s the end of my first week of actual development work, and I&#8217;m
happy with my progress. My <a href="https://github.com/csw/bioruby-maf/issues?milestone=1&amp;state=closed">first milestone</a>, for batch MAF
conversion to <a href="http://en.wikipedia.org/wiki/FASTA_format">FASTA</a>, is complete. I started with a simple
line-based parser, and then spent some time developing a faster one;
my MAF parser now appears to be quite competitive in performance. I
can now convert simple MAF files to FASTA with output matching that
from the <a href="http://galaxy.psu.edu/">Galaxy</a> tools. And I&#8217;m fairly confident in its
correctness, with RSpec <a href="https://github.com/csw/bioruby-maf/blob/master/spec/bio/maf/parser_spec.rb">unit tests</a> and Cucumber <a href="https://github.com/csw/bioruby-maf/blob/master/features/maf-parsing.feature">features</a>
specifying its behavior relatively completely.</p>

<h2>Parser development</h2>

<p>My initial approach was to simply process the file line-by-line,
calling <a href="http://www.ruby-doc.org/core-1.9.3/IO.html#method-i-readline">IO#readline</a> each time. This turned out to be a little more
cumbersome than I expected. Also, there was a <a href="http://lists.open-bio.org/pipermail/biopython-dev/2012-April/009561.html">discussion</a> on the
Biopython list of the usefulness of <a href="http://blastedbio.blogspot.com/2011/11/bgzf-blocked-bigger-better-gzip.html">BGZF</a> blocked compression for
MAF files, which would not fit well with such a linear line-oriented
approach. Inspired by Dmitry Vyukov&#8217;s <a href="http://www.1024cores.net/home/scalable-architecture/wide-finder-2">Wide Finder 2 entry</a>, I wrote
a <a href="https://github.com/csw/bioruby-maf/blob/d424245c047e3939d1e42b97c002c02fc8b578ac/lib/bio/maf/parser.rb#L61">new parser</a> which reads the file in 8 MB chunks. It immediately
locates the start of the last MAF alignment block in that chunk, and
then proceeds to scan through the chunk, parsing an alignment block at
a time, until the last block is reached. When it reaches the last
block, it <a href="https://github.com/csw/bioruby-maf/blob/d424245c047e3939d1e42b97c002c02fc8b578ac/lib/bio/maf/parser.rb#L130">joins the block fragments</a> at the end of one
chunk and the beginning of the next chunk.</p>

<p>For separating the alignment blocks within each chunk, using a
StringScanner and regexes worked well, but for parsing the individual
alignment blocks, using String#split to split the block into lines,
dispatching on the first character with a case statement and plain
string comparisons, and splitting lines into fields turned out to
perform substantially better. I suspect the overhead of setting up a
regex matching operation is the reason for this; profiling indicated
that it was spending considerable time in regex operations, at any rate.</p>

<p>After this optimization, the chunked parser was able to parse a 315
MB file and count its MAF blocks in 10.1 s, compared to 16.0 s for the
earlier line-based parser and 22.7 s for the parser from Galaxy&#8217;s
bx-python library. This seems like very respectable performance.</p>

<p>Also, the upcoming JRuby 1.7 on Java 7 appears to have excellent
performance with the chunked parser, perhaps due to the new
<a href="http://blog.headius.com/2011/08/jruby-and-java-7-what-to-expect.html">invokedynamic support</a>. After warming up the JVM, it averaged 16
&mu;s per alignment block parsed, compared to 25 &mu;s for Ruby
1.9.3. See the <a href="https://github.com/csw/bioruby-maf/wiki/Performance">performance page</a> on the project wiki for details.</p>

<h2>Behavior-Driven Development</h2>

<p>At the beginning of this week, I was not at all clear on when it would
make sense to use Cucumber and when to use RSpec. After writing a
<a href="https://github.com/csw/bioruby-maf/blob/d77ef2e133edf449a4ac3b18ff7d21052c0fc14a/features/maf-parsing.feature">Cucumber feature</a> to ensure that sequence data was exposed
properly, it didn&#8217;t seem worthwhile to duplicate that in
RSpec. However, the chunked parser was much harder to get right than
the original line-based parser, and RSpec was an invaluable tool for
that.</p>

<p>For most of the bugs I found, I developed an RSpec specification to
call for the correct behavior. Some of these could go into Cucumber
just as well, such as the one stipulating that a certain alignment
block should have ten sequences in it. Most, though, are about
low-level details such as correct parsing of
<a href="https://github.com/csw/bioruby-maf/blob/b3b9e060c6ba40bf17930eb6564615eb3b5b31bf/spec/bio/maf/parser_spec.rb#L138">sequence lines split across chunk boundaries</a>.</p>

<p>In some cases it was challenging to follow the BDD principle of
writing only the necessary code to make the next test pass. When I
started on the chunked parser, I had a very clear idea of how it
needed to work, and next to no idea how to specify that in terms of
tests. So I just started writing code and relying on the existing
parser tests. However, the set of RSpec tests I ended up with seems to
specify the chunked parser&#8217;s behavior reasonably well; they specify
where the last block in a chunk should be detected, how many blocks
should be parsed in a file if chunk fragments are being joined
properly, what happens if a sequence line is split across a chunk
boundary, and so on.</p>

<p>This gives me a concrete idea of the kind of tests I might want to
write to specify something like this next time. We&#8217;ll see how well I
can apply this next week, when I start working on indexed access.</p>

<h2>Tools</h2>

<p>I knew that it was possible to set up <a href="http://travis-ci.org/#!/csw/bioruby-maf">continuous integration</a> as a
hosted service with <a href="http://travis-ci.org/">Travis-CI</a>, but it was a pleasant surprise to
find that I could also have <a href="http://rubydoc.info/github/csw/bioruby-maf/">documentation</a> generated the same
way. It would be nice to have Travis-CI do hosted coverage reporting
as well, but setting up <a href="https://github.com/colszowka/simplecov">simplecov</a> was very easy.</p>

<p>I&#8217;ve been doing my development work with Emacs and using
<a href="http://barelyenough.org/projects/rspec-mode/">rspec-mode</a>, <a href="https://github.com/michaelklishin/cucumber.el">cucumber.el</a>, and <a href="https://github.com/magit/magit">Magit</a>. These have all been
better than nothing, but also somewhat frustrating; I&#8217;ve already sent
in one <a href="https://github.com/michaelklishin/cucumber.el/pull/21">pull request</a> for cucumber.el, and over the next
month or two I may explore some more radical ideas to make rspec-mode
and cucumber.el more useful. I&#8217;ve been gradually learning how to use
Magit, although I usually end up doing my more complex Git operations
from the command line.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/bioruby-maf/blog/2012/05/21/basic-maf-parsing/">Basic MAF Parsing</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2012-05-21T13:43:00-04:00" pubdate data-updated="true">May 21<span>st</span>, 2012</time>
        
         | <a href="/bioruby-maf/blog/2012/05/21/basic-maf-parsing/#disqus_thread">Comments</a>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Today is the official start of coding for GSoC. Following my
<a href="https://github.com/csw/bioruby-maf/wiki/Original-proposal">initial plan</a>,
I&#8217;m going to begin by implementing the basics of MAF parsing in
Ruby. I&#8217;ve already got Cucumber features defined for
<a href="https://github.com/csw/bioruby-maf/blob/master/features/maf-parsing.feature">basic MAF parsing</a>
and
<a href="https://github.com/csw/bioruby-maf/blob/master/features/maf-to-fasta.feature">conversion to FASTA</a>,
along with
<a href="https://github.com/csw/bioruby-maf/tree/master/test/data">reference data</a>
as processed by bx-python.</p>

<p>This should be fairly straightforward to implement, but I&#8217;m going to
wait until I&#8217;ve got this basic functionality running before I define
anything else in detail. I&#8217;m still weighing whether it will make sense
to have a &#8216;raw&#8217; representation of sequences and alignment blocks
rather than just using (and presumably subclassing) the
<a href="https://github.com/pjotrp/bioruby-alignment">bio-alignment</a>
representations.</p>

<p>Once basic MAF parsing works, my next area to focus on will be indexed
access. Many use cases for MAF involve pulling a few alignments out of
very large data files, rather than batch-processing the whole
file. I&#8217;ll be focusing on the indexed-access API at first, and
building a simple interim indexing scheme similar to that used by
Biopython, probably using SQLite in a similar way. In developing the
API, I&#8217;ll study those provided by bx-python and Biopython, the two
other MAF implementations providing persistent indexing.</p>

<p>Ultimately, I plan to revisit my actual indexing method, and
potentially implement support for bx-python&#8217;s
<a href="https://bitbucket.org/james_taylor/bx-python/src/07aca5a9f6fc/lib/bx/interval_index_file.py">interval index files</a>. I&#8217;ll
also take a careful look at other database alternatives such as
Berkeley DB and Tokyo Cabinet.</p>

<p>P.S. For my next blog post, I think I will try using Markdown&#8217;s
<a href="http://daringfireball.net/projects/markdown/syntax#link">reference-style links</a>,
since the raw source for these posts is getting unwieldy with inline
links.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/bioruby-maf/blog/2012/05/21/week_2_progress/">GSoC Week 2</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2012-05-21T11:13:00-04:00" pubdate data-updated="true">May 21<span>st</span>, 2012</time>
        
         | <a href="/bioruby-maf/blog/2012/05/21/week_2_progress/#disqus_thread">Comments</a>
        
      </p>
    
  </header>


  <div class="entry-content"><p>This was my second week of work on my GSoC project, and the last week
of the &#8216;community bonding&#8217; period before the official start of
coding. A major focus of mine was BioRuby&#8217;s
<a href="http://phyloxml.org/">phyloXML</a> support; it uses libxml, which has
been causing unit test failures under JRuby. In the end, the best
course of action seemed to separate the phyloXML support as a separate
plugin, which I have done as the
<a href="https://github.com/csw/bioruby-phyloxml">bio-phyloxml</a> gem. This will
remove BioRuby&#8217;s dependency on XML libraries entirely and that JRuby
issue along with it. At the same time, users of the phyloXML code
should be able to continue using it with no substantive changes.</p>

<p>Separately, I began porting this phyloXML code to use
<a href="http://nokogiri.org/">Nokogiri</a> instead of libxml-ruby, but ran into
difficulties with this effort. While it is possible, and the library
APIs are very similar, the code uses relatively low-level XML
processing APIs in ways that seem to be sensitive to subtle
differences in text node and namespace semantics between the two
libraries. Substantial restructuring of the code and the addition of
quite a few unit tests might be necessary to carry out such a port
with confidence that the resulting code would work well.</p>

<p>Also, someone else submitted a
<a href="https://github.com/jruby/jruby/pull/176">JRuby patch</a> for
<a href="http://jira.codehaus.org/browse/JRUBY-6658">JRUBY-6658</a>, one of the
major causes of BioRuby&#8217;s unit test failures with JRuby; once a fix is
integrated, we&#8217;ll be close to having all the tests passing under
JRuby.</p>

<p>I identified another JRuby bug,
<a href="http://jira.codehaus.org/browse/JRUBY-6666">JRUBY-6666</a>, causing
several unit test failures. This one affects BioRuby&#8217;s code for
running external commands, so it would be likely to be encountered in
production use. For this one, I also worked up a
<a href="https://github.com/jruby/jruby/pull/173">patch</a>.</p>

<p>I also spent some time preparing a performance testing environment,
for evaluating existing MAF implementations as well as my own. This
will be important, since I will be considering the use of an existing
C parser. I will also want to ensure that the performance of my code
is competitive with the alternatives. Lacking any hardware more
powerful than a MacBook Air, I am setting this up with Amazon EC2. To
simplify environment setup, I&#8217;ll be using
<a href="http://wiki.opscode.com/display/chef/Home">Chef</a>. I&#8217;ve already set up
a <a href="https://github.com/csw/chef-repo">Chef repository</a> with
configuration logic, and some
<a href="https://github.com/csw/ec2-launcher">rudimentary code</a> to streamline
launching Ubuntu machines on EC2 and bootstrapping a Chef
environment. To save money, I plan to make use of
<a href="http://aws.amazon.com/ec2/spot-instances/">EC2 Spot Instances</a>, which
are perfect for instances that only need to run for a few hours for
batch tasks.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/bioruby-maf/blog/2012/05/13/progress/">GSoC Week 1</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2012-05-13T18:14:00-04:00" pubdate data-updated="true">May 13<span>th</span>, 2012</time>
        
         | <a href="/bioruby-maf/blog/2012/05/13/progress/#disqus_thread">Comments</a>
        
      </p>
    
  </header>


  <div class="entry-content"><p>This has been my first half-week of work on my Google Summer of Code
project, and it&#8217;s off to an exciting start. The first order of
business has been to get my development environment together; since
I&#8217;ve been a microbiology student instead of a programmer for the last
year, it&#8217;s taken some work. In that process, I&#8217;ve ended up making a
few <a href="https://github.com/michaelklishin/cucumber.el/pull/21">open</a>
<a href="https://github.com/nibrahim/Hyde/pull/6">source</a>
<a href="https://github.com/nibrahim/Hyde/pull/7">contributions</a> just to get
my tools working the way I want. I&#8217;m running GNU Emacs 24 and trying
to take more advantage of it than I have in the past. I&#8217;ll have much
more to say about this in a future post.</p>

<p>I&#8217;ve also started working on the BioRuby unit test
<a href="http://travis-ci.org/#!/bioruby/bioruby/jobs/1296529">failures</a> under
JRuby, as a way of familiarizing myself with the BioRuby code base as
well as the community and its development processes. Right now, JRuby
in 1.8 mode is showing 6 failures and 126 errors, which is hardly
confidence-inspiring for people considering using JRuby with
BioRuby. This is too bad, since JRuby has some definite advantages as
a Ruby implementation. After looking into these failures, I&#8217;ve broken them
down into a few categories:</p>

<ul>
<li>temporary file permissions problems, likely due to some sort of Travis-CI
environment issue</li>
<li>a bug in JRuby&#8217;s implementation of
<a href="http://ruby-doc.org/stdlib-1.8.7/libdoc/open3/rdoc/Open3.html#method-c-popen3">Open3.popen3</a>
which I&#8217;m working up a bug report for</li>
<li>an odd autoload problem I&#8217;ve filed
<a href="https://jira.codehaus.org/browse/JRUBY-6658">JRUBY-6658</a> for and
sent an accompanying
<a href="https://github.com/rubyspec/rubyspec/pull/136">RubySpec patch</a>
for</li>
<li>a problem with libxml-jruby, which appears unmaintained, for which
I&#8217;ve submitted a
<a href="https://github.com/bioruby/bioruby/pull/55">BioRuby patch</a> plus
<a href="http://jira.codehaus.org/browse/JRUBY-6662">JRUBY-6662</a></li>
<li>and a small test case bug relating to floating point handling,
which I&#8217;ve submitted a
<a href="https://github.com/bioruby/bioruby/pull/54">patch</a> for.</li>
</ul>


<p>Once these are resolved, JRuby should be passing the BioRuby unit
tests in 1.8 mode, and closer to passing in 1.9 mode. (There are a few
extra failures under 1.9 that I haven&#8217;t sorted through yet.)</p>

<p>I&#8217;ve also gotten a start on my project itself, creating the
<a href="https://github.com/csw/bioruby-maf">bioruby-maf</a> Github repository
with a project skeleton and writing my
<a href="https://github.com/csw/bioruby-maf/blob/79004f9b75c1e33f9b265a1a97241d3c9d382997/features/maf-to-fasta.feature">first Cucumber feature</a>
for it. This is, in fact, my first Cucumber feature ever. However, I
did spend a few cross-country flights reading the
<a href="http://pragprog.com/book/achbd/the-rspec-book">RSpec</a> and
<a href="http://pragprog.com/book/hwcuc/the-cucumber-book">Cucumber</a> books
last week; between that and cribbing from
<a href="https://github.com/pjotrp/bioruby-alignment/tree/master/features">Pjotr&#8217;s code</a>
I feel like I have some idea what I&#8217;m doing. Just assembling that
feature has been useful, too, since I&#8217;ve had to get several of the
existing MAF tools running on my machine. In fact, my test MAF data and
the FASTA version of it are courtesy of
<a href="https://bitbucket.org/james_taylor/bx-python/wiki/Home">bx-python</a>,
which will be my reference implementation in many respects.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/bioruby-maf/blog/2012/05/09/hello-world/">Hello World</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2012-05-09T19:33:00-04:00" pubdate data-updated="true">May 9<span>th</span>, 2012</time>
        
         | <a href="/bioruby-maf/blog/2012/05/09/hello-world/#disqus_thread">Comments</a>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Hello World! This blog will be tracking the development of <a href="https://github.com/csw/bioruby-maf">bio-maf</a>, a <a href="https://cgwb.nci.nih.gov/FAQ/FAQformat.html#format5">Multiple Alignment Format</a> (MAF) parser for the Ruby bioinformatics library <a href="http://bioruby.open-bio.org/">BioRuby</a>, as part of the Google Summer of Code 2012. I&#8217;m Clayton Wheeler, a programmer and a biology student at the University of Michigan. This is exciting for me as my first substantial open source project, and as a way to write hopefully-useful bioinformatics software. And thanks, Google, for making this possible.</p>

<p>In my next post I&#8217;ll discuss what MAF is, what it&#8217;s useful for, and how I&#8217;m planning to approach the project. I&#8217;ll be making weekly status posts, plus others as inspiration strikes.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
    <a href="/bioruby-maf/blog/archives">Blog Archives</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/bioruby-maf/blog/2012/05/25/first_milestone/">First milestone</a>
      </li>
    
      <li class="post">
        <a href="/bioruby-maf/blog/2012/05/21/basic-maf-parsing/">Basic MAF parsing</a>
      </li>
    
      <li class="post">
        <a href="/bioruby-maf/blog/2012/05/21/week_2_progress/">GSoC Week 2</a>
      </li>
    
      <li class="post">
        <a href="/bioruby-maf/blog/2012/05/13/progress/">GSoC Week 1</a>
      </li>
    
      <li class="post">
        <a href="/bioruby-maf/blog/2012/05/09/hello-world/">Hello World</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>GitHub Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/csw">@csw</a> on GitHub
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'csw',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/bioruby-maf/javascripts/github.js" type="text/javascript"> </script>
</section>






  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Clayton Wheeler -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'biorubymafblog';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>







  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
