
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>First milestone - BioRuby MAF blog</title>
  <meta name="author" content="Clayton Wheeler">

  
  <meta name="description" content="It&#8217;s the end of my first week of actual development work, and I&#8217;m
happy with my progress. My first milestone, for batch MAF
conversion to &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://csw.github.com/bioruby-maf/blog/2012/05/25/first_milestone">
  <link href="/bioruby-maf/favicon.png" rel="icon">
  <link href="/bioruby-maf/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/bioruby-maf/javascripts/modernizr-2.0.js"></script>
  <script src="/bioruby-maf/javascripts/ender.js"></script>
  <script src="/bioruby-maf/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/bioruby-maf/atom.xml" rel="alternate" title="BioRuby MAF blog" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/bioruby-maf/">BioRuby MAF blog</a></h1>
  
    <h2>Multiple Alignment Format support for BioRuby and bio-alignment</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/bioruby-maf/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:csw.github.com/bioruby-maf" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/bioruby-maf/">Blog</a></li>
  <li><a href="/bioruby-maf/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div>
<article class="hentry" role="article">
  
  <header>
    
      <h1 class="entry-title">First Milestone</h1>
    
    
      <p class="meta">
        








  


<time datetime="2012-05-25T14:13:00-04:00" pubdate data-updated="true">May 25<span>th</span>, 2012</time>
        
         | <a href="#disqus_thread">Comments</a>
        
      </p>
    
  </header>


<div class="entry-content"><p>It&#8217;s the end of my first week of actual development work, and I&#8217;m
happy with my progress. My <a href="https://github.com/csw/bioruby-maf/issues?milestone=1&amp;state=closed">first milestone</a>, for batch MAF
conversion to <a href="http://en.wikipedia.org/wiki/FASTA_format">FASTA</a>, is complete. I started with a simple
line-based parser, and then spent some time developing a faster one;
my MAF parser now appears to be quite competitive in performance. I
can now convert simple MAF files to FASTA with output matching that
from the <a href="http://galaxy.psu.edu/">Galaxy</a> tools. And I&#8217;m fairly confident in its
correctness, with RSpec <a href="https://github.com/csw/bioruby-maf/blob/master/spec/bio/maf/parser_spec.rb">unit tests</a> and Cucumber <a href="https://github.com/csw/bioruby-maf/blob/master/features/maf-parsing.feature">features</a>
specifying its behavior relatively completely.</p>

<h2>Parser development</h2>

<p>My initial approach was to simply process the file line-by-line,
calling <a href="http://www.ruby-doc.org/core-1.9.3/IO.html#method-i-readline">IO#readline</a> each time. This turned out to be a little more
cumbersome than I expected. Also, there was a <a href="http://lists.open-bio.org/pipermail/biopython-dev/2012-April/009561.html">discussion</a> on the
Biopython list of the usefulness of <a href="http://blastedbio.blogspot.com/2011/11/bgzf-blocked-bigger-better-gzip.html">BGZF</a> blocked compression for
MAF files, which would not fit well with such a linear line-oriented
approach. Inspired by Dmitry Vyukov&#8217;s <a href="http://www.1024cores.net/home/scalable-architecture/wide-finder-2">Wide Finder 2 entry</a>, I wrote
a <a href="https://github.com/csw/bioruby-maf/blob/d424245c047e3939d1e42b97c002c02fc8b578ac/lib/bio/maf/parser.rb#L61">new parser</a> which reads the file in 8 MB chunks. It immediately
locates the start of the last MAF alignment block in that chunk, and
then proceeds to scan through the chunk, parsing an alignment block at
a time, until the last block is reached. When it reaches the last
block, it <a href="https://github.com/csw/bioruby-maf/blob/d424245c047e3939d1e42b97c002c02fc8b578ac/lib/bio/maf/parser.rb#L130">joins the block fragments</a> at the end of one
chunk and the beginning of the next chunk.</p>

<p>For separating the alignment blocks within each chunk, using a
StringScanner and regexes worked well, but for parsing the individual
alignment blocks, using String#split to split the block into lines,
dispatching on the first character with a case statement and plain
string comparisons, and splitting lines into fields turned out to
perform substantially better. I suspect the overhead of setting up a
regex matching operation is the reason for this; profiling indicated
that it was spending considerable time in regex operations, at any rate.</p>

<p>After this optimization, the chunked parser was able to parse a 315
MB file and count its MAF blocks in 10.1 s, compared to 16.0 s for the
earlier line-based parser and 22.7 s for the parser from Galaxy&#8217;s
bx-python library. This seems like very respectable performance.</p>

<p>Also, the upcoming JRuby 1.7 on Java 7 appears to have excellent
performance with the chunked parser, perhaps due to the new
<a href="http://blog.headius.com/2011/08/jruby-and-java-7-what-to-expect.html">invokedynamic support</a>. After warming up the JVM, it averaged 16
&mu;s per alignment block parsed, compared to 25 &mu;s for Ruby
1.9.3. See the <a href="https://github.com/csw/bioruby-maf/wiki/Performance">performance page</a> on the project wiki for details.</p>

<h2>Behavior-Driven Development</h2>

<p>At the beginning of this week, I was not at all clear on when it would
make sense to use Cucumber and when to use RSpec. After writing a
<a href="https://github.com/csw/bioruby-maf/blob/d77ef2e133edf449a4ac3b18ff7d21052c0fc14a/features/maf-parsing.feature">Cucumber feature</a> to ensure that sequence data was exposed
properly, it didn&#8217;t seem worthwhile to duplicate that in
RSpec. However, the chunked parser was much harder to get right than
the original line-based parser, and RSpec was an invaluable tool for
that.</p>

<p>For most of the bugs I found, I developed an RSpec specification to
call for the correct behavior. Some of these could go into Cucumber
just as well, such as the one stipulating that a certain alignment
block should have ten sequences in it. Most, though, are about
low-level details such as correct parsing of
<a href="https://github.com/csw/bioruby-maf/blob/b3b9e060c6ba40bf17930eb6564615eb3b5b31bf/spec/bio/maf/parser_spec.rb#L138">sequence lines split across chunk boundaries</a>.</p>

<p>In some cases it was challenging to follow the BDD principle of
writing only the necessary code to make the next test pass. When I
started on the chunked parser, I had a very clear idea of how it
needed to work, and next to no idea how to specify that in terms of
tests. So I just started writing code and relying on the existing
parser tests. However, the set of RSpec tests I ended up with seems to
specify the chunked parser&#8217;s behavior reasonably well; they specify
where the last block in a chunk should be detected, how many blocks
should be parsed in a file if chunk fragments are being joined
properly, what happens if a sequence line is split across a chunk
boundary, and so on.</p>

<p>This gives me a concrete idea of the kind of tests I might want to
write to specify something like this next time. We&#8217;ll see how well I
can apply this next week, when I start working on indexed access.</p>

<h2>Tools</h2>

<p>I knew that it was possible to set up <a href="http://travis-ci.org/#!/csw/bioruby-maf">continuous integration</a> as a
hosted service with <a href="http://travis-ci.org/">Travis-CI</a>, but it was a pleasant surprise to
find that I could also have <a href="http://rubydoc.info/github/csw/bioruby-maf/">documentation</a> generated the same
way. It would be nice to have Travis-CI do hosted coverage reporting
as well, but setting up <a href="https://github.com/colszowka/simplecov">simplecov</a> was very easy.</p>

<p>I&#8217;ve been doing my development work with Emacs and using
<a href="http://barelyenough.org/projects/rspec-mode/">rspec-mode</a>, <a href="https://github.com/michaelklishin/cucumber.el">cucumber.el</a>, and <a href="https://github.com/magit/magit">Magit</a>. These have all been
better than nothing, but also somewhat frustrating; I&#8217;ve already sent
in one <a href="https://github.com/michaelklishin/cucumber.el/pull/21">pull request</a> for cucumber.el, and over the next
month or two I may explore some more radical ideas to make rspec-mode
and cucumber.el more useful. I&#8217;ve been gradually learning how to use
Magit, although I usually end up doing my more complex Git operations
from the command line.</p>
</div>


  <footer>
    <p class="meta">
      
  

<span class="byline author vcard">Posted by <span class="fn">Clayton Wheeler</span></span>

      








  


<time datetime="2012-05-25T14:13:00-04:00" pubdate data-updated="true">May 25<span>th</span>, 2012</time>
      


    </p>
    
      <div class="sharing">
  
  <a href="http://twitter.com/share" class="twitter-share-button" data-url="http://csw.github.com/bioruby-maf/blog/2012/05/25/first_milestone/" data-via="" data-counturl="http://csw.github.com/bioruby-maf/blog/2012/05/25/first_milestone/" >Tweet</a>
  
  
  
</div>

    
    <p class="meta">
      
        <a class="basic-alignment left" href="/bioruby-maf/blog/2012/05/21/basic-maf-parsing/" title="Previous Post: Basic MAF parsing">&laquo; Basic MAF parsing</a>
      
      
        <a class="basic-alignment right" href="/bioruby-maf/blog/2012/06/04/indexed_maf_access/" title="Next Post: Indexed MAF access">Indexed MAF access &raquo;</a>
      
    </p>
  </footer>
</article>

  <section>
    <h1>Comments</h1>
    <div id="disqus_thread" aria-live="polite"><noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
</div>
  </section>

</div>

<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/bioruby-maf/blog/2012/08/05/bgzf_and_testing/">BGZF and testing</a>
      </li>
    
      <li class="post">
        <a href="/bioruby-maf/blog/2012/07/18/bio-maf_0.3.0/">Bio-MAF 0.3.0</a>
      </li>
    
      <li class="post">
        <a href="/bioruby-maf/blog/2012/07/09/bio-maf_0.2.0/">Bio-MAF 0.2.0</a>
      </li>
    
      <li class="post">
        <a href="/bioruby-maf/blog/2012/07/02/kyoto_cabinet_support_for_jruby/">Kyoto Cabinet support for JRuby</a>
      </li>
    
      <li class="post">
        <a href="/bioruby-maf/blog/2012/06/21/parallel_io/">Parallel I/O</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>GitHub Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating...</li>
  </ul>
  
  <a href="https://github.com/csw">@csw</a> on GitHub
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'csw',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/bioruby-maf/javascripts/github.js" type="text/javascript"> </script>
</section>






  
</aside>


    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Clayton Wheeler -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'biorubymafblog';
      
        
        // var disqus_developer = 1;
        var disqus_identifier = 'http://csw.github.com/bioruby-maf/blog/2012/05/25/first_milestone/';
        var disqus_url = 'http://csw.github.com/bioruby-maf/blog/2012/05/25/first_milestone/';
        var disqus_script = 'embed.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>







  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
